src/protocol/ipa/mod.rs

```diff
@@ -0,0 +1,207 @@
+use std::iter::{repeat, zip};
+
+use crate::{
+    error::Error,
+    ff::Field,
+    protocol::{
+        attribution::{
+            accumulate_credit::accumulate_credit, aggregate_credit::aggregate_credit,
+            credit_capping::credit_capping, AttributionInputRow,
+        },
+        context::Context,
+        sort::{
+            apply_sort::apply_sort_permutation,
+            generate_permutation::generate_permutation_and_reveal_shuffled,
+        },
+        RecordId,
+    },
+    secret_sharing::{Replicated, XorReplicated},
+};
+use futures::future::try_join_all;
+
+use super::modulus_conversion::{convert_all_bits, convert_all_bits_local};
+use super::{attribution::AggregateCreditOutputRow, context::SemiHonestContext};
+use crate::protocol::boolean::bitwise_equal::bitwise_equal;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    ModulusConversionForMatchKeys,
+    GenSortPermutationFromMatchKeys,
+    ApplySortPermutation,
+    ComputeHelperBits,
+    AccumulateCredit,
+    PerformUserCapping,
+    AggregateCredit,
+}
+
+impl crate::protocol::Substep for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::ModulusConversionForMatchKeys => "mod_conv_match_key",
+            Self::GenSortPermutationFromMatchKeys => "gen_sort_permutation_from_match_keys",
+            Self::ApplySortPermutation => "apply_sort_permutation",
+            Self::ComputeHelperBits => "compute_helper_bits",
+            Self::AccumulateCredit => "accumulate_credit",
+            Self::PerformUserCapping => "user_capping",
+            Self::AggregateCredit => "aggregate_credit",
+        }
+    }
+}
+
+/// # Errors
+/// Propagates errors from multiplications
+#[allow(dead_code)]
+pub async fn ipa<F>(
+    ctx: SemiHonestContext<'_, F>,
+    mk_shares: &[XorReplicated],
+    num_bits: u32,
+    other_inputs: Vec<Vec<Replicated<F>>>,
+    per_user_credit_cap: u32,
+) -> Result<Vec<AggregateCreditOutputRow<F>>, Error>
+where
+    F: Field,
+{
+    debug_assert_eq!(mk_shares.len(), other_inputs.len());
+
+    let local_lists = convert_all_bits_local(ctx.role(), mk_shares, num_bits);
```

Comment:
TODO: decrypt.

Reply:
I'm not sure what you mean.

Reply:
If we want to do the whole protocol, we'll need to decrypt shares.